### DISCUSSION TITLE
Add ability to set custom legend for an expression

### DISCUSSION THREAD

**Discussion by nitzanharel** [grafana/grafana#50202] [Dashboards, Panels & Visualizations]
For example - a query from Prometheus can return with labels/properties. You can select a specific label/property by setting {{label}} in the  legend field.
When you create an expression that is dependent on a query (e.g. $A * 2) - you get the entire legend (like setting the legend to "Auto") and you can't select a specific label/property

**Comment by wizardfiction** [comment]
Weirdly, sometimes my expression gets the appropriate name, and other times it doesn't.

Example:

`$A / ($B * $C)`, this gives me all of the data from $A along with the name I assigned to my expression, when I just want the expression name.

<img width="948" alt="Screen Shot 2022-09-12 at 9 16 59 PM" src="https://user-images.githubusercontent.com/17627585/189786136-aef73192-4029-45ec-aa6c-639e6147b84a.png">

When I do something like `$A / $B` instead, I get the correct name without all the extra stuff.

Having a hard time understanding why this is the case

**Comment by B3DTech** [comment]
You can use math directly in a prometheus query: 
(homeassistant_sensor_temperature_celsius{entity="sensor.main_floor_temperature"} * 1.8) + 32

**Comment by samlaf** [comment]
+1. This would be super useful to have.

**Comment by libbkmz** [comment]
In case of queries from the same datasource that would work. However, In my case, I have prometheus and graphite. How can it be possible to override the legend name for math expressions in grafana?

**Comment by r10r** [comment]
I also have several expressions with a long legend in a tiny font that are unreadable . What about using the expression name in the legend?

**Comment by itsmylife** [comment]
Thank you for submitting a feature request. We’ve reviewed your request and have decided not to implement this feature for the following reasons:
- Field override feature can be used to achieve the same result. Please check the documentation https://grafana.com/docs/grafana/latest/panels-visualizations/configure-overrides/

Even when we make the decision not to pursue a feature, the submission encourages conversations and ideas within our development team to continue making improvements, so please keep sharing!

**Reply by jdheyburn** [reply]
@itsmylife It would be helpful if you could provide an example. I only see Standard options > Display name to change the name of a series from an expression, and this does not support labels - setting `{{LABEL_NAME}}` prints it out literally.

**Reply by itsmylife** [reply]
@jdheyburn Please find the documentation about `Display Name`
https://grafana.com/docs/grafana/latest/panels-visualizations/configure-standard-options/#display-name

Example `${__field.labels.<the_label_name_inside_the_field>}`

**Reply by oddstr13** [reply]
It would be really nice to have the Options->Label for expressions too, if only for consistency's sake.

The above is an acceptable workaround, but having to search for the solution and use a different syntax in a different spot, is, in my humble opinion bad UX.

I am unlikely to remember this exact syntax the next time I need it, and have to look it up again.

**Comment by alicederyn** [comment]
In case anyone else hits this and wonders what on earth is being talked about, you want something like this:
![image](https://github.com/grafana/grafana/assets/979369/32849f0f-7b78-44c2-b70b-28a8ea271e90)
On the graph properties on the right, switch from "all" to "overrides" tab, then click "add field override", and select "fields returned by query". You can then select the query label for the expression you want (in this case, my expression was the fifth one, so called "E"). Then click "Add override property" and select "Display name", then type the label you want (probably not "foo"!).
![image](https://github.com/grafana/grafana/assets/979369/f74ae2e5-f381-4dab-b643-3cd44ccf0e43)

This is of course hard to find, hard to maintain, and deeply surprising. Overrides should obviously be listed front-and-center with the expression, not buried off to the right somewhere. But that's how to do it as the UI is currently laid out.

**Reply by jaymzh** [reply]
and you can't use variables in it :( So if you set it to "p99 {{some_label}}" it doesn't work.

**Reply by eugeniyk** [reply]
That's exactly the case - without variables the usage is quite limited

**Reply by rrileyca** [reply]
Hello @alicederyn, can you elaborate on where this setting is? I am using Grafana `v11.3.0-pre` and don't see this option at all. Our UI's look different too. For me when I edit a panel, on the far right (where the Panel Options are), the overrides are at the bottom of the list. I don't see an "All/Overrides" tab. Further, when I add an override, I do not have an option for "Display Name". Thanks!

Edit: Actually, I can find "Display Name" as an override option but only on single-datasource dashboards. The one mixed dashboard I have does not support this feature.

**Reply by eferroni** [reply]
> and you can't use variables in it :( So if you set it to "p99 {{some_label}}" it doesn't work.

actually, you can add a variable using:
`${__field.labels.<name_of_your_field>}`
